test/contract-access-controller-integration.test.js

```diff
@@ -0,0 +1,226 @@
+        identity: id2
+      })
+    })
+
+    after(async () => {
+      if (orbitdb1) {
+        await orbitdb1.stop()
+      }
+
+      if (orbitdb2) {
+        await orbitdb2.stop()
+      }
+
+      if (ipfsd1) {
+        await stopIpfs(ipfsd1)
+      }
+
+      if (ipfsd2) {
+        await stopIpfs(ipfsd2)
+      }
+    })
+
+    describe('OrbitDB Integration', function () {
+      let db, db2
+      let dag
+      let dbManifest, acManifest
+
+      before(async () => {
+        db = await orbitdb1.feed('AABB', {
+          identity: id1,
+          accessController: {
+            type: 'eth-contract',
+            web3: web3,
+            abi: abi,
+            contractAddress: contract._address
+          }
+        })
+
+        db2 = await orbitdb2.feed(db.address, { identity: id2, accessController: { web3: web3 } })
+        await db2.load()
+
+        dag = await ipfs1.object.get(db.address.root)
+        dbManifest = JSON.parse(dag.toJSON().data)
+        const hash = dbManifest.accessController.split('/').pop()
+        const acManifestDag = await ipfs1.object.get(hash)
+        acManifest = JSON.parse(acManifestDag.toJSON().data)
+      })
+
+      it('makes database use the correct access controller', async () => {
+        assert.equal(acManifest.params.contractAddress, db.access.contract._address)
+      })
+
+      it('saves database manifest file locally', async () => {
+        assert.notEqual(dbManifest, null)
+      })
+
+      it('saves access controller manifest file locally', async () => {
+        assert.notEqual(acManifest, null)
+      })
+
+      describe('database manifest', () => {
+        it('has correct name', async () => {
+          assert.equal(dbManifest.name, 'AABB')
+        })
+
+        it('has correct type', async () => {
+          assert.equal(dbManifest.type, 'feed')
+        })
+
+        it('has correct address', async () => {
+          assert.notEqual(dbManifest.accessController, null)
+          assert.equal(dbManifest.accessController.indexOf('/ipfs'), 0)
+        })
+      })
+
+      describe('access controller manifest', () => {
+        it('has correct type', async () => {
+          assert.equal(acManifest.type, 'eth-contract')
+        })
+
+        it('has correct address', async () => {
+          assert.equal(acManifest.params.contractAddress.indexOf('0x'), 0)
+          assert.equal(acManifest.params.contractAddress, contract._address)
+        })
+      })
+
+      describe('access controls', () => {
+        it('granting access enables to write to the database', async () => {
+          let err
+          try {
+            await db.add('hello?')
+            assert.equal('Should not end here', false)
+          } catch (e) {
+            err = e
+          }
+
+          assert.equal(err, `Error: Could not append entry, key "${id1.id}" is not allowed to write to the log`)
+          await db.access.grant('write', id1.id, { from: accounts[0] })
+          await db.add('hello!')
```

Comment:
We don't need this add here as it gets called below in the try/catch block, right?

Reply:
First one should throw an error, but separated them to different tests :+1: